### PR TITLE
sdl1/sessions

### LINKED ISSUE #3: Session management for python API wrapper
Following https://github.com/ScottishCovidResponse/SimulationData.jl/pull/2 we need to make it possible to use session management with the python API. I think this can be done with `@pywith`, I have an initial proof-of-concept and it seems to work. So we would end up being able to do something like
```julia
SimpleNetworkSimAPI("../data_pipeline_api/examples/test_data_2/config.yaml") do api
    df1 = read_table(api, "human/mixing-matrix", "abc")
    df2 = read_table(...)
    ...
 end # allows the python api to do its __exit__ method or whatever
```

### PR DESCRIPTION
**Pull request by sdl1**
- Allows using the API in a session
- Does some parsing into Julia objects

example (this is using https://github.com/ScottishCovidResponse/data_pipeline_api/pull/14)
```julia
julia> cfg = "test/data/config.yaml"
julia> StandardAPI(cfg) do api
           df = read_table(api, "object", "example-table")
           @show df
           @show typeof(df)
           dist = read_distribution(api, "parameter", "example-distribution")
           @show dist
           @show typeof(dist)
       end
[PYTHON]: Reading table
df = 2×2 DataFrame
│ Row │ a     │ b     │
│     │ Int64 │ Int64 │
├─────┼───────┼───────┤
│ 1   │ 1     │ 3     │
│ 2   │ 2     │ 4     │
typeof(df) = DataFrames.DataFrame
[PYTHON]: Reading distribution
dist = Distributions.Gamma{Float64}(α=1.0, θ=2.0)
typeof(dist) = Distributions.Gamma{Float64}
Distributions.Gamma{Float64}
``` 

the resulting `access-example.yaml` file:
```yaml
data_directory: .
open_timestamp: 2020-06-30 11:12:30.668067
close_timestamp: 2020-06-30 11:12:32.105128
io:
- type: read
  timestamp: 2020-06-30 11:12:30.673237
  call_metadata:
    data_product: object
    component: example-table
  access_metadata:
    data_product: object
    component: example-table
    version: 1.0.0
    filename: object/example.h5
    run_id: example
    calculated_hash: ec3d8ffc77b2c2e25e2050b9ef309699ef4d9989
- type: read
  timestamp: 2020-06-30 11:12:31.868742
  call_metadata:
    data_product: parameter
    component: example-distribution
  access_metadata:
    data_product: parameter
    component: example-distribution
    version: 1.0.0
    filename: parameter/example.toml
    run_id: example
    calculated_hash: ef068f657c7fa07deeb4b7f1f0845737d5894117
```

### CODE REVIEW

File: src/pyapi.jl [hunk 1]
```diff
@@ -5,73 +5,101 @@
...
 end
 
+struct SimpleNetworkSimAPI <: FileAPI
+    pyapi::PyObject
+end
 
-function read_estimate(data_product, component)
-    d = py"read_estimate($data_product, $component)"
-    return d
+function StandardAPI(f::Function, config_filename)
```
Comment (jwscook) [inline src/pyapi.jl:96]: This is where the magic happens so a few comments here would be helpful.

**Comment by jwscook** [comment]
This looks great!

I have a question about `src/pyapi.jl`: `StandardAPI` is imported in a `py""" """` block, and declared as a `struct`. I assume that the python version is contained in the scope of the `py""" """` block and is only accessed from calls to python?

File: src/pyapi.jl (hunk 1 above)
Comment (sdl1) [reply src/pyapi.jl:96]: Yep, I added some docstrings and comments.

**Comment by sdl1** [comment]
> I have a question about src/pyapi.jl: StandardAPI is imported in a py""" """ block, and declared as a struct. I assume that the python version is contained in the scope of the py""" """ block and is only accessed from calls to python?

Yes, my understanding of PyCall is only basic, but the intention is:
- `StandardAPI` inside the `py""` refers to the python class of that name, from `data_pipeline_api`
- Additionally we define a Julia struct of the same name to wrap this python object (not strictly necessary, but makes things cleaner and allows dispatch if we need it)
- The two should never get confused, because the scopes are separate

**Comment by AlexRobson** [comment]
This looks good. Was able to reproduce all the tests locally. If there are no further issues, will merge in tomorrow a.m. 
Need to update the README too but that can be a follow up MR as it's not in the changeset.

File: src/pyapi.jl [hunk 2]
```diff
@@ -5,73 +5,154 @@
...
+    df = read_table(api, ...)
+    ...
+end
+```
+
+## Returns
+- The value returned by `f`
+"""
+function StandardAPI(f::Function, config_filename)
+    result = nothing
```
Comment (AlexRobson) [inline src/pyapi.jl:97]:
```suggestion
    result = nothing # Required
```

**Review by AlexRobson — APPROVED** [review]

File: Project.toml [hunk 3]
```diff
@@ -6,6 +6,7 @@ version = "0.1.1"
 [deps]
```
Comment (AlexRobson) [inline Project.toml:6]: bump major version

Comment (sdl1) [reply Project.toml:6]: Done

**Pull request merged**